### ISSUE TITLE
Measuring exploitability for large games

### ISSUE THREAD

**Issue by jeremysalwen** [deepmind/open_spiel#772]
I'm training nfsp agents on hearts using `leduc_nfsp.py`.  However, I was forced to disable the nash_conv computation, since it appears to be trying to compute the complete set of possible information states in the game which blowing up my memory/cpu.

Is there a sampling based implementation that could efficiently calculate approximate nash_conv for games with large game trees?  or is there another exploitability metric I should be using instead?

**Comment by lanctot** [comment]
Hi @jeremysalwen, at the moment you can try using RL to exploit your opponent, see the [rl_response](https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/examples/rl_response.py) example.

We have all the components to be able to implement [ABR](https://arxiv.org/abs/2004.09677). It would require some programming and could be a nice project, if you're interested, but it's non-trivial. We could help, though.. and it'd make a great contribution!

**Comment by jeremysalwen** [comment]
Interesting, I'll have to read the ABR paper.  rl_response seems like it might be too expensive to be running repeatedly during training, maybe I would save that until training is complete? Is there something more lightweight I could run to just compare performance to earlier agents to check progress, like alpha-rank?

**Comment by lanctot** [comment]
ABR is even more expensive because it does search too. Typically we do by saving checkpoints and evaluating them separately.

Yes you can compute the payoff tensor between checkpoints and then do some kind of analysis of the progression. Alpha-rank is a good choice, you can also do uniform distribution and measure head-to-head performance. We came up with a few others too (variants of Nash averaging frim Balduzzi et al), all with various properties. We are posting a paper to arXiv on this topic in the next week or so, I will post the link here when it's live.

**Comment by jeremysalwen** [comment]
I'm looking for examples in the repository where fixed agents are evaluated against each other by playing a sample of games.  I'm having trouble finding something that seems appropriate for the NFSP agents generated by leduc_nfsp.py.

1. `evaluate_bots.py` has the evaluated_bots function which seems to evaluate the performance of a set of bots playing a game.  However, it seems like the "bot" interface is totally separate than the RL agent interface, and so there is no way to use this function for the RL agents generated by NFSP.
2. `expected_game_score.py` seems to evaluate based on `policy` objects, which we are generating in NFSP, but it seems to assume that the complete game tree is tractable, and so is inappropriate for larger games.
3. `egt/alpharank.py` seems to assume I have already computed the self play data in some existing format.  e.g. `alpharank_example.py` just calls `xfp_solver.get_empirical_metagame()` to compute this metagame date using code specific to the `XFPSolver`. I was expecting there would be some generic tool to compute this metagame data for any set of agents.

I know I could implement some script specific to NFSP to evaluate agents the way I want, but this seems wasteful of effort.  Is there some more standard way of doing this that I am missing?

**Comment by lanctot** [comment]
Hi @jeremysalwen, you are not missing anything. We indeed do not have an example of that anywhere. It would be useful to have one, even if just for NFSP agents. If you do this, please contribute an example to the repo for the next person :)

**Comment by jeremysalwen** [comment]
Ok, that makes sense.  Could you provide some guidance for what interface I should be using? Should I be trying to make the RL agents conform to the bot interface? Should I be operating on policies? Should I be operating on AbstractAgents?

**Comment by lanctot** [comment]
@jeremysalwen, my apologies.. somehow I completely missed your last message.

I think the value we'd have by having this example in any form outweighs any small detriment to it being under a specific choice of API or usage. So I'm a bit indifferent. But, if I had to choose, I'd say the RL Agent API would be the best one. That is: it would be great to have something that could define a set of RLAgents that are loaded and run (in evaluation mode) to build the matrix of values, and then we could compute something from that matrix (be it AlphaRank or Nash Averaging or something). 

It would then be wonderful to see an example use of this run over previous NFSP agent checkpoints, showing in a two-player zero-sum game that the Nash Averaging solution puts most weight on the final iterates. Or, if you're doing multiplayer like hearts, then maybe the AlphaRank distibution.

**Comment by jeremysalwen** [comment]
I ended up implementing a subset of this functionality, just calculating average returns. (It seems like the alpharank code requires a dense metagame matrix, which makes it complicated, since I am just sampling a few games).  The code is just a part of the leduc_nfsp.py example, but I tried to write it in a way that it could be reasonably generalized.

Before I clean up/document everything for the PR, I just want to make sure, is this something that is worthwhile, that you would want in open_spiel?

https://github.com/deepmind/open_spiel/pull/793
